Integrate strimzi-kafka-oauth 0.6.0 features (#3336)
strimzi/strimzi-kafka-operator: documentation/modules/oauth/con-oauth-reauthentication.adoc

// Module included in the following assemblies:
//
// assembly-oauth.adoc

[id='con-oauth-reauthentication-{context}']

> = Session re-authentication mechanism

Reviewer: ```suggestion
= Session re-authentication for Kafka brokers
```

> The Kafka _SASL OAUTHBEARER_ mechanism supports a so called 're-authentication' mechanism.

Reviewer: ```suggestion
The Kafka _SASL OAUTHBEARER_ mechanism supports a Kafka feature called the _re-authentication_ mechanism.
```

Reviewer: _Add a new sentence to the same paragraph:_

OAuth 2.0 authentication in Strimzi provides support for the re-authentication mechanism.

> When enabled (by setting `maxSecondsWithoutReauthentication` attribute of 'oauth' type `authentication` listener configuration), the authenticated session expires when the access token expires.

Reviewer: ```suggestion
When the re-authentication mechanism is enabled in the configuration of an `oauth` type listener, the broker's authenticated session expires when the access token expires.
```

> 

Reviewer: _Make line 12 part of the above paragraph because I think it's strongly related._

Author: AFAICT the line 12 ('The client must ...') is already part of the paragraph that starts with 'When the re-authentication mechanism ...'

> The client then has to re-authenticate the existing session by sending a fresh valid access token to the Kafka broker, without dropping the connection.

Reviewer: ```suggestion
The client must then re-authenticate to the existing session by sending a new, valid access token to the broker, without dropping the connection.
```

> If token validation is successful a new client session is started over the existing connection.

Reviewer: ```suggestion
If token validation is successful, a new client session is started using the existing connection.
```

> If the client fails to re-authenticate, any further attempts to use the connection for sending or receiving messages results in Kafka broker closing the connection.

Reviewer: ```suggestion
If the client fails to re-authenticate, the broker will close the connection if further attempts are made to send or receive messages.
```

> Java clients using Kafka client library 2.2 or later will automatically re-authanticate if re-authentication is enabled on the broker.

Reviewer: ```suggestion
Java clients that use Kafka client library 2.2 or later automatically re-authenticate if the re-authentication mechanism is enabled on the broker.
```

Reviewer: I suggest making line 16 a separate paragraph. It feels like a standalone point.

> 

Reviewer: _New paragraph to explain the "how to":_

You enable session re-authentication in the `Kafka` resource. Set the `maxSecondsWithoutReauthentication` property for a TLS listener with `type: oauth` authentication.

> Re-authentication mechanism was introduced in Kafka version 2.2. You can read more about it in https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate[KIP-368].

Reviewer: ```suggestion
For more information about the re-authentication mechanism, which was added in Kafka version 2.2, see link:https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate[KIP-368^].
```

Reviewer: _Please include cross-refs to three related modules._

_Here is the Asciidoc markup:_

```
.Additional resources

* xref:con-oauth-authentication-broker-{context}[]

* xref:proc-oauth-authentication-broker-config-{context}[]

* xref:appendix_crds#type-KafkaListenerAuthenticationOAuth-reference[KafkaListenerAuthenticationOAuth schema reference]
```

// Module included in the following assemblies:
//
// assembly-oauth.adoc

[id='con-oauth-reauthentication-{context}']
= Session re-authentication for Kafka brokers

> The Kafka _SASL OAUTHBEARER_ mechanism supports a Kafka feature called the _re-authentication_ mechanism.

Reviewer: ```suggestion
The Kafka _SASL OAUTHBEARER_ mechanism, which is used for OAuth 2.0 authentication in Strimzi, supports a Kafka feature called the _re-authentication_ mechanism.
```

> OAuth 2.0 authentication in Strimzi provides support for the re-authentication mechanism.

Reviewer: Delete line 9.

I think that simplifies the introduction a bit!

When the re-authentication mechanism is enabled in the configuration of an `oauth` type listener, the broker's authenticated session expires when the access token expires.
The client must then re-authenticate to the existing session by sending a new, valid access token to the broker, without dropping the connection.

If token validation is successful, a new client session is started using the existing connection.
If the client fails to re-authenticate, the broker will close the connection if further attempts are made to send or receive messages.
Java clients that use Kafka client library 2.2 or later automatically re-authenticate if the re-authentication mechanism is enabled on the broker.

You enable session re-authentication in the `Kafka` resource. Set the `maxSecondsWithoutReauthentication` property for a TLS listener with `type: oauth` authentication.

> 

Reviewer: _Suggested addition to the above paragraph:_

Session re-authentication is supported for both types of token validation (fast local JWT and introspection endpoint). For an example configuration, see xref:proc-oauth-authentication-broker-config-{context}[].
